### ISSUE TITLE
[BUG] internal compiler error: in assign_temp, at function.c:977

### ISSUE THREAD

**Issue by MarcDirven** [google/benchmark#986]
**Describe the bug**
I have created a lazy evaluated string splitter. Its iterator `operator*` returns a reference to a sub-`std::string` (if `C++14`). When I want to iterate over its substrings and add a `benchmark::DoNotOptimize(substring)`, I get the following error: 
```
during RTL pass: expand
In file included from benchmarks.cpp:1:
include/benchmark/benchmark.h: In function 'void BM_StringSplitter(benchmark::State&)':
include/benchmark/benchmark.h:322:48: internal compiler error: in assign_temp, at function.c:977
   asm volatile("" : "+m,r"(value) : : "memory");
                                                ^
libbacktrace could not find executable to open
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://sourceforge.net/projects/mingw-w64> for instructions.
mingw32-make.exe[3]: *** [bench\CMakeFiles\Benchmark.dir\build.make:63: bench/CMakeFiles/Benchmark.dir/benchmarks.cpp.obj] Error 1
mingw32-make.exe[2]: *** [CMakeFiles\Makefile2:1292: bench/CMakeFiles/Benchmark.dir/all] Error 2
mingw32-make.exe[1]: *** [CMakeFiles\Makefile2:1299: bench/CMakeFiles/Benchmark.dir/rule] Error 2
mingw32-make.exe: *** [Makefile:631: Benchmark] Error 2
```
Now this is when it gets weird. I have no problems whatsoever when iterating in a non-benchmark situation. I am aware however, that this is a compiler bug, so I'm not 100% sure where to report the issue.

The function looks as follows:
```cpp
static void BM_StringSplitter(benchmark::State& state) {
    std::string toSplit = "Benchmark Benchmark Benchmark Benchmark Benchmark "
                          "Benchmark Benchmark Benchmark Benchmark Benchmark "
                          "Benchmark Benchmark Benchmark Benchmark Benchmark "
                          "Benchmark Benchmark Benchmark Benchmark Benchmark "
                          "Benchmark Benchmark Benchmark Benchmark Benchmark "
                          "Benchmark Benchmark Benchmark Benchmark Benchmark "
                          "Benchmark Benchmark Benchmark Benchmark Benchmark "
                          "Benchmark Benchmark Benchmark Benchmark Benchmark ";

    for (auto _ : state) {
        auto splitter = lz::split(toSplit, " ");
        benchmark::DoNotOptimize(splitter);

        for (std::string& substring : splitter) {
            benchmark::DoNotOptimize(substring); // This causes the error
        }
    }
}
```

**System**
Which OS, compiler, and compiler version are you using:
  - OS: Windows 10 10.0.18363 Build 18363
  - Compiler and version: mingw-w64 x86_64-8.1.0-posix-seh-rt_v6-rev0 / w64 6.0

**To reproduce**
1. sync to commit https://github.com/MarcDirven/cpp-lazy (I have just pushed the issue to my own repository)

Basically, everything is set up except the google benchmark itself, since that's in my .gitignore file. Please run the benchmark target and/or see the `BM_StringSplitter` function in `cpp-lazy/bench/benchmarks.cpp`, line 57.

**Expected behavior**
Either a workaround or a fix. I really want to benchmark this function.

**Additional context**
Not sure what to give for additional context, but if there are any question then by all means, ask me.

**Comment by dominichamon** [comment]
I believe this is a mingw compiler issue. I don't have the OS/compiler pair to test.

Can you try replacing the `std::string&` with a `const std::string&` in the loop? it shouldn't change anything but i'm curious :)

**Comment by MarcDirven** [comment]
> 
> 
> I believe this is a mingw compiler issue. I don't have the OS/compiler pair to test.
> 
> Can you try replacing the `std::string&` with a `const std::string&` in the loop? it shouldn't change anything but i'm curious :)

Actually, that díd fix it, probably because the `DoNotOptimize` function is overloaded with a `const` variant? I don't know, I'm not very familiar with the library. Thanks for the alternative.

**Comment by dominichamon** [comment]
in the non-const version (and non-clang) which is the branch you're hitting, we specify that value is an output as read-write memory or a general purpose register. In the const version we're specifying that there is no output but value can be input as register or memory.

i _think_ we need a MINGW specific version for the non-const that has some other asm form.

@EricWF ?

**Comment by MarcDirven** [comment]
Not that my input matters much since I am not very familiar with X86_64 instructions, so I don't know from which compiler version of GCC it happens. At least from mingw-w64 x86_64-8.1.0-posix-seh-rt_v6-rev0 / w64 6.0

**Comment by KirillLykov** [comment]
I have the same problem with `g++ (GCC) 7.3.1 20180303 ` on RedHat Linux.

**Comment by imrichardcole** [comment]
@MarcDirven I'm taking a look into the backlog involving Windows related issues.  I think your repo has long since moved on.  Would you be able to confirm which version of cpp-lazy you were trying to benchmark at the time please?

**Comment by MarcDirven** [comment]
I didn't do much versioning back then but you can use commit 02a8b98cb83e7e4141ba8b2d7edbcde0f9a016c4 (https://github.com/MarcDirven/cpp-lazy/commit/02a8b98cb83e7e4141ba8b2d7edbcde0f9a016c4)

**Comment by imrichardcole** [comment]
Thanks @MarcDirven.  Using my local build of HEAD from benchmark, it works without issue:

![image](https://github.com/user-attachments/assets/578b7af0-249d-4f36-b494-15cdf45df6d9)

Sum of changes which seem unrelated to the original issue posted:

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index f325bb7..1cac5d5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -41,8 +41,13 @@ target_include_directories("cpp-lazy" INTERFACE
         ${CMAKE_CURRENT_LIST_DIR}/include
         )

+set(benchmark_DIR "C:/sandbox/work/oss/benchmark_main/install/lib/cmake/benchmark")
+find_package(benchmark
+        REQUIRED
+        CONFIG
+)
+
 if(CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
     add_subdirectory(tests)
-    add_subdirectory(extern/benchmark)
     add_subdirectory(bench)
 endif()
\ No newline at end of file
diff --git a/bench/benchmarks.cpp b/bench/benchmarks.cpp
index 3bb9357..b762fa0 100644
--- a/bench/benchmarks.cpp
+++ b/bench/benchmarks.cpp
@@ -68,14 +68,8 @@ static void BM_StringSplitter(benchmark::State& state) {
         auto splitter = lz::split(toSplit, " ");
         benchmark::DoNotOptimize(splitter);

-//#if __cplusplus < 201703L || (defined(_MSVC_LANG) && _MSVC_LANG < 201703L)
-//        using StringType = std::string&;
-//#else
-//        using StringType = std::string_view;
-//#endif
-
         for (std::string& substring : splitter) {
-            benchmark::DoNotOptimize(substring);
+            benchmark::DoNotOptimize(substring); // This causes the error
         }
     }
 }
@@ -85,7 +79,7 @@ static void BM_TakeWhile(benchmark::State& state) {
     std::array<int, size> array = lz::range(static_cast<int>(size)).toArray<size>();

     for (auto _ : state) {
-        auto takeWhile = lz::takewhile(array, [](const int i) { return i != size - 1; });
+        auto takeWhile = lz::takewhile(array, [&](const int i) { return i != size - 1; });
         benchmark::DoNotOptimize(takeWhile);

         for (int taken : takeWhile) {
```

**Comment by MarcDirven** [comment]
@imrichardcole Well if that mingw version doesn't throw the error, then maybe it has been implicitly fixed?

Back then I used mingw compiler but prefer clang/msvc nowadays so reproducing requires a tiny bit more effort. I guess if reproducing isn't possible anymore then this issue can be closed.

**Comment by LebedevRI** [comment]
These kinds of issues usually crop up in early release candidates versions of compilers,
and usually are fixed in later compiler versions, so when that happens,
it really is best to report that upstream (to the compiler issue tracker).

**Comment by imrichardcole** [comment]
Apologies my tests were done with MSVC.   Let me try with mingw and report back
